pkg/controller/apiextensions/definer/reconciler.go

```diff
@@ -0,0 +1,275 @@
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+	"github.com/crossplane/crossplane-runtime/pkg/resource/unstructured"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v1alpha1"
+	"github.com/crossplane/crossplane/pkg/controller/apiextensions/composite"
+)
+
+const (
+	// TODO(muvaf): consider adding _shorterWait_ for non-error cases where the result
+	// is expected to be quick. For example, deletion or CRD establishment takes
+	// unnecessarily long just because shortWait is 10 seconds.
+	shortWait      = 10 * time.Second
+	longWait       = 1 * time.Minute
+	timeout        = 2 * time.Minute
+	maxConcurrency = 5
+	finalizerName  = "finalizer.apiextensions.crossplane.io"
+
+	errGetInfraDef           = "cannot get infrastructure definition"
+	errGetCRD                = "cannot get crd"
+	errApplyCRD              = "cannot apply the generated crd"
+	errUpdateInfraDefStatus  = "cannot update status of infrastructure definition"
+	errCannotStartController = "cannot start controller"
+	errStopController        = "cannot stop controller"
+	errAddFinalizer          = "cannot add finalizer"
+	errRemoveFinalizer       = "cannot remove finalizer"
+	errDeleteCRD             = "cannot delete crd"
+
+	waitingInstanceDeletion   = "waiting for all cr instances to be gone"
+	waitingCRDEstablish       = "waiting for crd to be established"
+	errNotControllerByDefiner = "cannot start a controller for a crd that this definer does not control"
+)
+
+// Setup adds a controller that reconciles ApplicationConfigurations.
+func Setup(mgr ctrl.Manager, log logging.Logger) error {
+	name := "apiextensions/" + strings.ToLower(v1alpha1.InfrastructureDefinitionGroupKind)
+	r := NewReconciler(mgr,
+		WithLogger(log),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	// TODO(muvaf): register this reconciler to events from CRDs, too.
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.InfrastructureDefinition{}).
+		WithOptions(controller.Options{MaxConcurrentReconciles: maxConcurrency}).
+		Complete(r)
+}
+
+// NewReconciler returns a new *reconciler.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) reconcile.Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+	newDefinerFn := func() Definer { return &v1alpha1.InfrastructureDefinition{} }
+	r := &reconciler{
+		client:     kube,
+		Finalizer:  resource.NewAPIFinalizer(kube, finalizerName),
+		crd:        NewAPIInfrastructureClient(kube),
+		mgr:        mgr,
+		newDefiner: newDefinerFn,
+		log:        logging.NewNopLogger(),
+		recorder:   event.NewNopRecorder(),
+	}
+
+	for _, f := range opts {
+		f(r)
+	}
+	// TODO(muvaf): we don't have a use case but it'd be consistent if we allowed
+	// different controller engines to be configured via ReconcilerOption.
+	r.ctrl = composite.NewControllerEngine(mgr, r.log)
```

Comment:
Like a lot of our options my guess is that it would help us inject testing dependencies, even if in the real world we'd pretty much always use `composite.NewControllerEngine`.